File: app/src/main/java/com/limelight/binding/input/driver/BleDriverService.java
package com.limelight.binding.input.driver;

import android.app.Service;
import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothManager;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Binder;
import android.os.IBinder;

import com.limelight.LimeLog;

import java.util.ArrayList;

public class BleDriverService extends Service implements UsbDriverListener {

    private BluetoothAdapter bluetoothAdapter;
    private boolean started;

    private final BleDriverBinder binder = new BleDriverBinder();
    private final ArrayList<AbstractController> controllers = new ArrayList<>();

    private UsbDriverListener listener;
    private int nextDeviceId = 100; // Start at 100 to avoid conflict with USB devices

    @Override
    public void reportControllerState(int controllerId, int buttonFlags, float leftStickX, float leftStickY,
                                      float rightStickX, float rightStickY, float leftTrigger, float rightTrigger) {
        if (listener != null) {
            listener.reportControllerState(controllerId, buttonFlags, leftStickX, leftStickY, rightStickX, rightStickY, leftTrigger, rightTrigger);
        }
    }

    @Override
    public void reportControllerMotion(int controllerId, byte motionType, float motionX, float motionY, float motionZ) {
        if (listener != null) {
            listener.reportControllerMotion(controllerId, motionType, motionX, motionY, motionZ);
        }
    }

    @Override
    public void deviceRemoved(AbstractController controller) {
        controllers.remove(controller);
        if (listener != null) {
            listener.deviceRemoved(controller);
        }
    }

    @Override
    public void deviceAdded(AbstractController controller) {
        if (listener != null) {
            listener.deviceAdded(controller);
        }
    }

    public class BleDriverBinder extends Binder {
        public void setListener(UsbDriverListener listener) {
            BleDriverService.this.listener = listener;
            if (listener != null) {
                for (AbstractController controller : controllers) {
                    listener.deviceAdded(controller);
                }
            }
        }

        public void start() {
            BleDriverService.this.start();
        }

        public void stop() {
            BleDriverService.this.stop();
        }
    }

    private void start() {
        if (started) return;

        SharedPreferences prefs = getSharedPreferences("ble_prefs", Context.MODE_PRIVATE);
        String mac = prefs.getString(BlePairingActivity.PREF_PAIRED_BLE_CONTROLLER, null);

        if (mac != null && bluetoothAdapter != null && bluetoothAdapter.isEnabled()) {
            BluetoothDevice device = bluetoothAdapter.getRemoteDevice(mac);
            LimeLog.info("BleDriverService attempting to connect to " + mac);
            ProConBleDriver controller = new ProConBleDriver(this, device, nextDeviceId++, this);
            if (controller.start()) {
                controllers.add(controller);
                started = true;
            }
        }
    }

    private void stop() {
        if (!started) return;
        started = false;

        while (controllers.size() > 0) {
            controllers.remove(0).stop();
        }
    }

    @Override
    public void onCreate() {
        super.onCreate();
        BluetoothManager bluetoothManager = (BluetoothManager) getSystemService(Context.BLUETOOTH_SERVICE);
        if (bluetoothManager != null) {
            this.bluetoothAdapter = bluetoothManager.getAdapter();
        }
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        LimeLog.info("BleDriverService: onStartCommand executed!");
        start();
        return START_STICKY;
    }

    @Override
    public void onDestroy() {
        stop();
        listener = null;
        super.onDestroy();
    }

    @Override
    public IBinder onBind(Intent intent) {
        return binder;
    }
}

File: app/src/main/java/com/limelight/binding/input/driver/BleDriverService.kt
package com.limelight.binding.input.driver

import android.app.Service
import android.content.Intent
import android.os.Binder
import android.os.IBinder
import com.limelight.LimeLog
import com.limelight.nvstream.jni.MoonBridge
import com.switch2.controllers.core.Switch2Controller
import com.switch2.controllers.core.Switch2ControllerListener
import com.switch2.controllers.core.Switch2ControllerState
import com.switch2.controllers.core.Switch2MotionState
import com.switch2.controllers.manager.Switch2Manager
import java.util.concurrent.ConcurrentHashMap

class BleDriverService : Service() {
    private val binder = BleDriverBinder()
    private var listener: UsbDriverListener? = null
    private lateinit var switch2Manager: Switch2Manager
    private val controllerWrappers = ConcurrentHashMap<Int, AbstractController>()

    private inner class Switch2ControllerWrapper(
        private val controller: Switch2Controller,
        listener: UsbDriverListener,
    ) : AbstractController(controller.controllerId, listener, controller.vendorId, controller.productId) {

        init {
            type = MoonBridge.LI_CTYPE_NINTENDO
            capabilities = (
                MoonBridge.LI_CCAP_GYRO.toInt() or
                MoonBridge.LI_CCAP_ACCEL.toInt() or
                MoonBridge.LI_CCAP_RUMBLE.toInt()
            ).toShort()
            supportedButtonFlags = controller.supportedButtonFlags
        }

        override fun start(): Boolean = true

        override fun stop() = Unit

        override fun rumble(lowFreqMotor: Short, highFreqMotor: Short) {
            switch2Manager.rumble(controller.controllerId, lowFreqMotor, highFreqMotor)
        }

        override fun rumbleTriggers(leftTrigger: Short, rightTrigger: Short) {
            switch2Manager.rumble(controller.controllerId, leftTrigger, rightTrigger)
        }
    }

    private val switch2Listener = object : Switch2ControllerListener {
        override fun onControllerAdded(controller: Switch2Controller) {
            LimeLog.info("BleDriverService: Switch2Controller added: ${controller.name} (id: ${controller.controllerId})")
            val currentListener = listener ?: return
            val wrapper = Switch2ControllerWrapper(controller, currentListener)
            controllerWrappers[controller.controllerId] = wrapper
            currentListener.deviceAdded(wrapper)
        }

        override fun onControllerRemoved(controller: Switch2Controller) {
            LimeLog.info("BleDriverService: Switch2Controller removed: ${controller.name} (id: ${controller.controllerId})")
            val wrapper = controllerWrappers.remove(controller.controllerId)
            val currentListener = listener ?: return
            if (wrapper != null) {
                currentListener.deviceRemoved(wrapper)
            }
        }

        override fun onControllerStateReported(state: Switch2ControllerState) {
            listener?.reportControllerState(
                state.controllerId,
                state.buttonFlags,
                state.leftStickX,
                state.leftStickY,
                state.rightStickX,
                state.rightStickY,
                state.leftTrigger,
                state.rightTrigger,
            )
        }

        override fun onControllerMotionReported(motion: Switch2MotionState) {
            listener?.reportControllerMotion(
                motion.controllerId,
                motion.motionType,
                motion.x,
                motion.y,
                motion.z,
            )
        }
    }

    override fun onCreate() {
        super.onCreate()
        switch2Manager = Switch2Manager.getInstance(this)
        switch2Manager.start()
    }

    override fun onBind(intent: Intent?): IBinder = binder

    inner class BleDriverBinder : Binder() {
        fun setListener(listener: UsbDriverListener?) {
            this@BleDriverService.listener = listener
            if (listener != null) {
                val currentControllers = switch2Manager.getConnectedControllers()
                LimeLog.info("BleDriverService: setListener, current connected controllers: ${currentControllers.size}")
                for (ctrl in currentControllers) {
                    val wrapper = Switch2ControllerWrapper(ctrl, listener)
                    controllerWrappers[ctrl.controllerId] = wrapper
                    listener.deviceAdded(wrapper)
                }
            } else {
                controllerWrappers.clear()
            }
        }

        fun start() {
            LimeLog.info("BleDriverService: start() connecting Switch2Manager")
            switch2Manager.start()
            switch2Manager.addListener(switch2Listener)
        }

        fun stop() {
            LimeLog.info("BleDriverService: stop()")
            switch2Manager.removeListener(switch2Listener)
            controllerWrappers.clear()
        }
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        switch2Manager.start()
        return START_STICKY
    }

    override fun onDestroy() {
        switch2Manager.removeListener(switch2Listener)
        controllerWrappers.clear()
        listener = null
        super.onDestroy()
    }
}
